fix(fetch): defer global access to avoid module-load TypeError

Jye10032 opened the pull request:
## Description

Fixes #7259

Similar to #7153 which fixed unsafe destructuring in `config.env`, this PR fixes another unsafe destructuring pattern in the fetch adapter that causes `TypeError` when `utils.global` is `undefined` during module loading.

## Problem

The current code (lines 15-17 in `lib/adapters/fetch.js`) directly destructures `utils.global` without checking if it's defined:

```javascript
const globalFetchAPI = (({Request, Response}) => ({
  Request, Response
}))(utils.global);
```

This throws `TypeError: Cannot destructure property 'Request' of 'undefined'` in certain build environments (Vite, webpack) where module initialization timing can cause `utils.global` to be temporarily undefined.

## Summary
Defer `utils.global` access to `factory()` and guard it with a fallback object to prevent module-
load crashes when globals are not ready.

## Fix
- Move `Request/Response/ReadableStream/TextEncoder` access into `factory()`
- Use a fallback object to avoid destructuring from `undefined`

## Reproduction
Minimal repro repo:
https://github.com/Jye10032/axios-repo

Steps:
1. `npm install`
2. `npm run build`
3. `npm run preview`
4. Open the preview URL → console shows:

TypeError: Cannot destructure property 'Request' of 'undefined'

## Related

- Similar fix in #7155 for `config.env` destructuring
- Reported by users experiencing build failures in production deployments

Jye10032 committed:
fix: add fallback for utils.global to prevent destructuring error

When utils.global is undefined (e.g., in certain bundler configurations),
destructuring Request, Response, ReadableStream, and TextEncoder will throw
a TypeError. This fix adds default empty object fallbacks to both
destructuring statements to prevent the error and allow axios to gracefully
fall back to the XHR adapter.

Fixes #7259

alibassam24 commented:
gg well played

Jye10032 committed:
fix(fetch): defer global destructuring

jasonsaayman reviewed (changes requested):
Thanks @Jye10032. When fetch.js's top-level code runs before `utils.js`'s default export has finished populating (a circular-import / evaluation-order quirk that bundlers like Vite can surface), `utils.global` reads as `undefined`, and the original destructure has no default, so it throws at module load before anything in axios can recover. Moving the read into `factory()` sidesteps that entirely; by the time `factory` is invoked from `getFetch()`, the ESM graph has settled.

Two small simplifications before this lands.

Flatten the fallback. `utils.global` is itself `globalThis ?? self ?? window ?? global`, so the only realistic case where it's undefined is the circular-import one and in that case `globalThis` is equally available. The `?? {}` arm is unreachable in any environment that runs ESM:

```js
const globalObject = utils.global ?? globalThis;
````

Drop the IIFE around Request/Response. With globalObject already in scope, the IIFE's only purpose (handling destructure-of-undefined at module load) is gone:

```js
const { Request, Response } = globalObject;
const globalFetchAPI = { Request, Response };
```

Same shape going into the `utils.merge.call(…, globalFetchAPI, env)` below.

Optional but nice: a regression test that stubs `utils.global` to undefined, calls `factory({})`, and asserts no throw. The bundler-graph case is hard to reproduce in unit tests directly, but stubbing the module is a reasonable proxy.

Jye10032 committed:
fix(fetch): defer global access in fetch adapter

Jye10032 commented:
Rebased onto the latest `v1.x` and updated the PR based on your feedback.

  Changes in the latest revision:
  - moved the `utils.global` read into `factory()`
  - simplified the fallback to `utils.global ?? globalThis`
  - removed the extra IIFE/default-destructure guard
  - added a regression test covering `utils.global = undefined` to ensure the fetch adapter no
  longer crashes during initialization

  I also re-ran the local fetch adapter tests and verified the original repro no longer throws the
  `Request` destructuring error.
